### ISSUE TITLE
Unexpected response body received

### ISSUE THREAD

**Issue by unknownuser19982** [Cornutum/tcases#266]
Hi, I was working with the autogenerate unit tests that tcases created, and when I was running them I ran into this trouble. And could not figure out why it the test case was failing. 

The following picture is the error produced by the failing testcase
![unexpected](https://user-images.githubusercontent.com/121258451/209209538-a9555357-1a57-44ad-bda2-bcaf8857040f.PNG)

And this is the printed response
![response](https://user-images.githubusercontent.com/121258451/209209676-091179f3-75f6-407f-a518-26cd0f9b2495.PNG)

If it helps, I am currently using JHipster and using the auto-generated code from jhipster

**Comment by kerrykimbrough** [comment]
If I understand your report correctly,  the test executed a `POST` request that contained a body with Content-Type `application/x-www-form-urlencoded`. But the request failed with status code 415 because this content type is not supported by the server. But the 415 response also contained a body with Content-Type `application/problem+json`. Consequently, response validation failed because `application/problem+json` was not the expected Content-Type. 

Does that explanation help? Or is there some reason why this `ResponseValidationException` is not correct?

**Comment by unknownuser19982** [comment]
Hmm. What would be the expected content-type? Would it be application/x-www-form-urlencoded?

**Comment by kerrykimbrough** [comment]
No, it looks like application/x-www-form-urlencoded was the **_request body_** content type that was refused. The issue is the content type expected for the **_response body_**. Your OpenAPI definition should specify that. 

It looks like the actual content type of the response was `application/problem+json`. Is that not what is specified in the OpenAPI definition?

**Comment by unknownuser19982** [comment]
So i set the content-type of the response ot be application/json. But it is still giving me the error
![application_json](https://user-images.githubusercontent.com/121258451/209357996-97563332-d0c9-44a9-8cf2-735d015a1c39.PNG)
Here is the response of the openapi spec
![openapispec](https://user-images.githubusercontent.com/121258451/209358072-8af6fb3c-057b-4dbd-9eba-2a501015e329.PNG)

**Comment by kerrykimbrough** [comment]
This OpenAPI definition of the response for status code 415 contains no `content` object definition. That means that a response with status code 415 is expected to contain no body. But that actual response _did_ contain a body. That's why you're seeing an "unexpected response body received" exception.

**Comment by unknownuser19982** [comment]
Hi, so I have removed the response body entirely, but I am still getting the same error
![no body](https://user-images.githubusercontent.com/121258451/209411480-7d903399-44da-4212-b502-9977eddde6e4.PNG)

**Comment by kerrykimbrough** [comment]
Sorry, there is not enough info in these screen shots for me to know what is going on. If you could send the *-Responses.json file, the result of the println() statements, and the full text of the resulting exception, I might be able to learn more.

**Comment by unknownuser19982** [comment]
Here is a screenshot of the error. The println is in under the debug console. The line with (3) is the last three print statements
![full_screenshot](https://user-images.githubusercontent.com/121258451/209413294-a5737b66-7311-46f2-a1fd-97e6c97f8f4e.PNG)

Here is the full error
`org.cornutum.tcases.openapi.test.ResponseValidationException: POST / (415), body: unexpected response body received
 at org.cornutum.tcases.openapi.test.ResponseValidator.assertBodyValid(ResponseValidator.java:141)
 at cantech.api.testcasegeneration.msimpl.MovieApiTest.post_BodyDefined_Is_No(MovieApiTest.java:114)
 at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
 at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
 at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.lang.reflect.Method.invoke(Method.java:498)
 at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
 at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
 at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
 at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
 at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
 at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
 at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
 at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
 at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
 at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
 at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
 at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
 at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
 at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
 at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
 at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
 at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
 at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)`

And here is the response-json file
```JSON
{
  "/" : {
    "get" : {
      "200" : {
        "description" : "Request was successful!",
        "content" : {
          "application/json" : {
            "schema" : {
              "maxItems" : 25,
              "minItems" : 0,
              "type" : "array",
              "example" : [ {
                "title" : "The Dark Knight",
                "director" : "Christopher Nolan",
                "description" : "When the menace known as the Joker wreaks havoc and chaos on the people of Gotham, Batman must accept one of the greatest psychological and physical tests of his ability to fight injustice. \n",
                "rating" : 4.5
              }, {
                "title" : "The Shawshank Redemption",
                "director" : "Frank Darabont",
                "description" : "Two imprisoned men bond over a number of years, finding solace and eventual redemption through acts of common decency.\n",
                "rating" : 5.0
              } ],
              "items" : {
                "required" : [ "description", "director", "rating", "title" ],
                "type" : "object",
                "properties" : {
                  "id" : {
                    "minLength" : 0,
                    "type" : "string",
                    "example" : "moviesAreDefinedByIdString"
                  },
                  "title" : {
                    "type" : "string"
                  },
                  "director" : {
                    "type" : "string"
                  },
                  "description" : {
                    "type" : "string"
                  },
                  "rating" : {
                    "maximum" : 5,
                    "minimum" : 0,
                    "type" : "number"
                  }
                },
                "example" : {
                  "title" : "The Dark Knight",
                  "director" : "Christopher Nolan",
                  "description" : "When the menace known as the Joker wreaks havoc and chaos on the people of Gotham, Batman must accept one of the greatest psychological and physical tests of his ability to fight injustice. \n",
                  "rating" : 4.5
                }
              }
            }
          }
        }
      }
    },
    "post" : {
      "201" : {
        "description" : "Successfully Created",
        "content" : {
          "application/json" : {
            "schema" : {
              "required" : [ "description", "director", "rating", "title" ],
              "type" : "object",
              "properties" : {
                "id" : {
                  "minLength" : 0,
                  "type" : "string",
                  "example" : "moviesAreDefinedByIdString"
                },
                "title" : {
                  "type" : "string"
                },
                "director" : {
                  "type" : "string"
                },
                "description" : {
                  "type" : "string"
                },
                "rating" : {
                  "maximum" : 5,
                  "minimum" : 0,
                  "type" : "number"
                }
              },
              "example" : {
                "title" : "The Dark Knight",
                "director" : "Christopher Nolan",
                "description" : "When the menace known as the Joker wreaks havoc and chaos on the people of Gotham, Batman must accept one of the greatest psychological and physical tests of his ability to fight injustice. \n",
                "rating" : 4.5
              }
            },
            "example" : {
              "id" : 101
            }
          }
        }
      },
      "415" : {
        "description" : "Bad Body formatting"
      },
      "400" : {
        "description" : "Bad Body formatting"
      },
      "405" : {
        "description" : "Bad Body formatting"
      }
    }
  }
}
```

**Comment by kerrykimbrough** [comment]
Thanks, I understand the problem now. `response.asString()` is returning an empty string, but `assertBodyValid` is assuming that "no body" will be represented by a null string.  In which case, the empty string looks like unexpected body content. 

Look for a new release soon that will fix this problem.

**Comment by unknownuser19982** [comment]
I see. Is there a way to make the this test case pass? Would changing the open API spec such that it can return a body for the 4XX responses be a way to make it pass this test case?

**Comment by kerrykimbrough** [comment]
This would probably avoid the problem.

```json
"415" : {
        "description" : "Bad Body formatting",
        "content": {
            "text/plain": {
             }
         }
      }
```

**Comment by unknownuser19982** [comment]
Hi so I was able to change the openAPI spec and was able to get through the test. However, I have encountered the following error 

`java.lang.AssertionError: 1 expectation failed.
Expected status code (a value equal to or greater than [400] and a value less than [500]) but a value equal to or greater than [400] [201] was less than [400].
 at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
 at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
 at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
 at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
 at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:80)
 at org.codehaus.groovy.reflection.CachedConstructor.doConstructorInvoke(CachedConstructor.java:74)
 at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrap.callConstructor(ConstructorSite.java:84)
 at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:59)
 at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:237)
 at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:249)
 at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure.validate(ResponseSpecificationImpl.groovy:494)
 at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure$validate$1.call(Unknown Source)
 at io.restassured.internal.ResponseSpecificationImpl.validateResponseIfRequired(ResponseSpecificationImpl.groovy:656)
 at sun.reflect.GeneratedMethodAccessor116.invoke(Unknown Source)
 at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.lang.reflect.Method.invoke(Method.java:498)
 at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
 at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:190)
 at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:58)
 at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:168)
 at io.restassured.internal.ResponseSpecificationImpl.statusCode(ResponseSpecificationImpl.groovy:125)
 at io.restassured.internal.ValidatableResponseOptionsImpl.statusCode(ValidatableResponseOptionsImpl.java:114)
 at cantech.api.testcasegeneration.msimpl.MovieApiTest.post_BodyApplicationJsonValuePropertiesDescriptionType_Is_NotString(MovieApiTest.java:384)
 at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
 at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
 at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.lang.reflect.Method.invoke(Method.java:498)
 at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
 at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
 at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
 at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
 at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
 at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
 at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
 at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
 at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
 at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
 at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
 at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
 at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
 at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
 at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
 at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
 at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
 at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)`

When this test was run
![newerror](https://user-images.githubusercontent.com/121258451/209755534-50f4e17d-2393-44b5-aeea-5ab76679bb7b.PNG)

The problem is that the server was suppose to send a response that the request was invalid. However, it actually registered the request as possible and processed the response. However, when I run the same body on postman, I get the proper response that the request was invalid. The following is the Postman response
![postman error](https://user-images.githubusercontent.com/121258451/209755754-a95cf492-8f73-47f0-8158-87a272f77b1b.PNG)

As you can see, there is an inconsistency in the test. Using the same body as the generated test in Postman yields the correct response, however, running the testcase does not yield the proper result. Furthermore, there were also some inconsistencies with what test was passing and what test was failing, I found that some test cases sometimes pass and sometimes fail, which is a bit non-deterministic.  The following screenshot is the whole generated testcases, there are test cases that are also testing for 'is_notString' where it is passing. Is this a potential new bug?
![testcases](https://user-images.githubusercontent.com/121258451/209756041-73f0b33b-ebfc-4254-a7c6-377d4d069743.PNG)

**Comment by kerrykimbrough** [comment]
This request posts a JSON object with a 'description' property that is supposed to be a string but instead is a number. That is technically in violation of the OpenAPI spec, which is why it should be a failure. But it looks like your API server is accepting it anyway. Is that a bug in your API? Or a bug in the OpenAPI spec?

**Comment by kerrykimbrough** [comment]
OK, I see that the Postman execution caused the expected failure.  

Perhaps there is a problem in the way the request is formed by the Tcases/RestAssured test. Can you capture the raw request data actually received by the server?

**Comment by unknownuser19982** [comment]
> This request posts a JSON object with a 'description' property that is supposed to be a string but instead is a number. That is technically in violation of the OpenAPI spec, which is why it should be a failure. But it looks like your API server is accepting it anyway. Is that a bug in your API? Or a bug in the OpenAPI spec?

Interestingly enough, I believe it may be caused by the way JHipster parses the request body. When the request body is 

```json
{
    "director":"director",
    "title":true,
    "description":"description",
    "rating":4.5
}
```
Notice that the type for `"title": true` is boolean but the required is `"title": String` . This might mean that when JHipster parses the body, it does not identify the 'true' as boolean but instead parses it as a string. Below the code is the generated response, note that the title gets converted to string
```json
{
    "title": "true",
    "director": "director",
    "description": "description",
    "rating": 4.5
}
```

**Comment by kerrykimbrough** [comment]
Fixed in release 4.0.0
